### PR TITLE
add quit command to cli_wallet #1050

### PR DESCRIPTION
**Pull request by nanomobile**
PR for #1050.

### CODE REVIEW

File: libraries/wallet/include/graphene/wallet/wallet.hpp [hunk 1]
```diff
@@ -632,6 +632,12 @@ class wallet_api
        */
       bool    load_wallet_file(string wallet_filename = "");
 
+      /** Exits from Graphene wallet.
+       *
+       * The current wallet will be closed.
+       */
+      void    exit_wallet();
+
```
Comment (oxarbitrage) [inline libraries/wallet/include/graphene/wallet/wallet.hpp]: please just name this function `exit` or `quit`. the user already is inside the wallet.

File: libraries/wallet/wallet.cpp [hunk 2]
```diff
@@ -770,6 +770,15 @@ class wallet_api_impl
 
       return true;
    }
+
+    void exit_wallet()
+    {
+        ilog( "exiting wallet ..." );
+        save_wallet_file();
+        ilog( "closed wallet successfully" );
+        exit(0);
```
Comment (oxarbitrage) [inline libraries/wallet/wallet.cpp]: i used `exit` before in other parts of the program and @pmconrad recommended to not use it. please check: https://github.com/bitshares/bitshares-core/pull/928#pullrequestreview-120729779

File: tests/cli/main.cpp [hunk 3]
```diff
@@ -388,3 +388,30 @@ BOOST_AUTO_TEST_CASE( cli_set_voting_proxy )
    }
    app1->shutdown();
 }
+
+///////////////
```
Comment (oxarbitrage) [inline tests/cli/main.cpp]: test cases for exit command dont need a test case, it is actually not testing anything here.

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: I've made changes in commit - just renamed exit_wallet command to quit name. Will check exit and change it too soon, just need to check & test more.

File: tests/cli/main.cpp (hunk 3 above)
Comment (nanomobile) [reply tests/cli/main.cpp]: in test case I check next command quit:

con.wallet_api_ptr->quit();

Comment (oxarbitrage) [reply tests/cli/main.cpp]: just executing the command does not mean it is testing anything.

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (oxarbitrage) [reply libraries/wallet/wallet.cpp]: you are going to need to send a new pull request or rebase, pull can have only your changes.

File: tests/cli/main.cpp (hunk 3 above)
Comment (nanomobile) [reply tests/cli/main.cpp]: what should be done to correctly testing quit command ? can you direct me in right way please ?

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: yes sure, just made mistake, sorry, will fix soon with rebase or new PR

File: tests/cli/main.cpp (hunk 3 above)
Comment (oxarbitrage) [reply tests/cli/main.cpp]: just remove the test case fully, not needed.

Comment (nanomobile) [reply tests/cli/main.cpp]: ok understood ! Thanks ! Just thought test case needed to check if there are any exceptions while quiting wallet

Comment (nanomobile) [reply tests/cli/main.cpp]: removed fully test case

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: fixed already with rebase

Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: changed also FC_THROW instead of exit

File: libraries/wallet/include/graphene/wallet/wallet.hpp (hunk 1 above)
Comment (nanomobile) [reply libraries/wallet/include/graphene/wallet/wallet.hpp]: quit name was choosed by me

File: libraries/wallet/wallet.cpp [hunk 4]
```diff
@@ -770,6 +770,17 @@ class wallet_api_impl
...
    }
+
+    void quit()
+    {
+        ilog( "exiting wallet ..." );
+        save_wallet_file();
+        ilog( "closed wallet successfully" );
+        
+        FC_THROW("closed wallet successfully");
+        throw;
```
[the remark this answers is not in the document]
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: even throw won't exit, need to find another solution, soon will add changes in commit for this PR

[the remark this answers is not in the document]
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: is it ok to use std::exit instead ?

File: libraries/wallet/include/graphene/wallet/wallet.hpp [hunk 5]
```diff
@@ -632,6 +632,12 @@ class wallet_api
        */
       bool    load_wallet_file(string wallet_filename = "");
 
+      /** Exits from Graphene wallet.
```
[the remark this answers is not in the document]
Comment (nanomobile) [reply libraries/wallet/include/graphene/wallet/wallet.hpp]: ok sure thanks

File: libraries/wallet/wallet.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/wallet/wallet.cpp]: `std:exit` and `exit` is the same. IMHO `exit` is fine here. `throw` is not appropriate as no error or unexpected situation here. i am not sure if we should not use `exit` anywhere, in this particular case to me is the best. 

another option should be to do it by `editline` if they have a workaround. not sure if this is possible. i will like to have opinion from @jmjatlanta on this as @pmconrad is on vacation.

rest looks good.

[the remark this answers is not in the document]
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: ok thank you very much !

File: libraries/wallet/wallet.cpp [hunk 6]
```diff
@@ -770,6 +770,15 @@ class wallet_api_impl
 
       return true;
    }
+
+    void quit()
+    {
+        ilog( "exiting wallet ..." );
+        save_wallet_file();
+        ilog( "closed wallet successfully" );
+        std::exit(0);
```
Comment (abitmore) [inline libraries/wallet/wallet.cpp]: I don't think it's good to call `exit` in a library.

File: libraries/wallet/wallet.cpp [hunk 7]
```diff
@@ -770,6 +770,15 @@ class wallet_api_impl
 
       return true;
    }
+
+    void quit()
+    {
+        ilog( "exiting wallet ..." );
+        save_wallet_file();
```
Comment (abitmore) [inline libraries/wallet/wallet.cpp]: What if this throws an exception?

File: libraries/wallet/wallet.cpp (hunk 6 above)
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: what is the best to call instead of exit in your opinion ? @abitmore can you help please with right direction ?

File: libraries/wallet/wallet.cpp (hunk 7 above)
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: will ad try/catch then, right ? do you mean that ?

File: libraries/wallet/wallet.cpp (hunk 6 above)
Comment (nanomobile) [reply libraries/wallet/wallet.cpp]: @abitmore sorry maybe I missunderstood you about exit in a library, but here in cli_wallet it's OK do you think ? right ? :-)

**Comment by nanomobile** [comment]
@ryanRfox @abitmore @oxarbitrage @jmjatlanta @clockworkgr I've added test case as requested and try/catch block, still cannot understand what to use instead of EXIT, can anybody advise anything what is the best to use instead of EXIT ?

Thanks everybody !

**Comment by oxarbitrage** [comment]
please research for an alternative and present for discussion all the options you can find, we don't know either.

**Comment by oxarbitrage** [comment]
the test is not enough in my opinion. it needs to create some accounts, move some balances, then `quit`. then open again, use the file to restore and check if the movements done are there.

**Comment by nanomobile** [comment]
ok sure ! Thanks ! :+1:

**Comment by nanomobile** [comment]
I see that even in libraries exit is still used in many different places, for example:
1. https://github.com/bitshares/bitshares-core/blob/master/libraries/app/application.cpp#L425
2. https://github.com/bitshares/bitshares-core/blob/master/libraries/app/application.cpp#L988
3. https://github.com/bitshares/bitshares-core/blob/master/libraries/app/application.cpp#L1019

and in the whole project there are also other places with exit method. So what can you say about that guys @oxarbitrage @abitmore @ryanRfox @jmjatlanta @clockworkgr ? Why there is it OK but not here ? Or there isn't OK too ?

**Comment by abitmore** [comment]
@nanomobile thanks for the report. We need to fix those. Exiting from a library makes it very hard if not impossible to test with boost testing framework. For the 2 exiting on error, should change to throw an exception. For the other one, probably should return a special value then check at where calling it. Need to create new issues for these though.

**Comment by nanomobile** [comment]
there are not only 3 places with exit method, we need to fix all IMHO

**Comment by abitmore** [comment]
Created new issue: https://github.com/bitshares/bitshares-core/issues/1110, please feel free to add more findings there. Thanks.

**Comment by cogutvalera** [comment]
just rebased current changes in this PR to be synced with core/develop, soon will make more commits in order to make everything correctly and close this issue

**Comment by cogutvalera** [comment]
instead of using exit(0) it's much better to use _Exit(0) as described here http://www.cplusplus.com/reference/cstdlib/_Exit/ :

> **Terminate calling process**
> Terminates the process normally by returning control to the host environment, but without performing any of the regular cleanup tasks for terminating processes (as function exit does).
> 
> No object destructors, nor functions registered by atexit or at_quick_exit are called.
> 
> Whether C streams are closed and/or flushed, and files open with tmpfile are removed depends on the particular system or library implementation.
> 
> If status is zero or EXIT_SUCCESS, a successful termination status is returned to the host environment.
> If status is EXIT_FAILURE, an unsuccessful termination status is returned to the host environment.
> Otherwise, the status returned depends on the system and library implementation.
> **Data races**
> Concurrently calling this function multiple times has no effect.
> 
> **Exceptions (C++)**
> No-throw guarantee: this function never throws exceptions.

**Comment by cogutvalera** [comment]
guys check and test this pull request please if this issue can be closed or not yet ...
Thanks !

**Comment by oxarbitrage** [comment]
i cant understand 2 things:

- why use http://www.cplusplus.com/reference/cstdlib/_Exit/ if it says it will make a non clean shut down? we want to exit in a clean way IMO.
- why the save of the wallet on `quit` was removed ? i think it was a good idea to have that.

many thanks.

**Comment by cogutvalera** [comment]
@oxarbitrage maybe you didn't understood my message correctly that's why:

1. _Exit terminates the process normally by returning control to the host environment, but without performing any of the regular cleanup tasks for terminating processes (as function exit does). So it terminates the process normally not the same way as exit does. Read please again more carefully http://www.cplusplus.com/reference/cstdlib/_Exit/ this doc.

2. When you call ^C the wallet file won't be save that's why I've decided to make the same for QUIT command, just to exit without saving wallet file because there maybe crashes as @abitmore reported earlier. So users must do backups and save their wallet files with special command save_wallet_file.

**Comment by oxarbitrage** [comment]
2- `control-c` exit the wallet without saving. `control-d` saves the wallet and exit. to avoid the need to save the wallet manually i think the behavior of `control-d` is preferred. when @abitmore mentioned that save could crash i think he was referring to `try/catch` the call as you did before at some point.

1- we want to avoid `exit()` and also `_exit()`. i see  no reason to `_exit` unclean, leaving connections open, orphan processes or anything like that if we can avoid it. 

ideally we want to exit by `throw` or `return`. i understand this can be tricky as the `main` function is here: https://github.com/bitshares/bitshares-core/blob/master/programs/cli_wallet/main.cpp
i am not sure how to do it but if we can send a signal from the library(when `quit` is executed) to the main and `return` there will be a good option to consider.

File: tests/cli/main.cpp [hunk 8]
```diff
@@ -387,4 +387,4 @@ BOOST_AUTO_TEST_CASE( cli_set_voting_proxy )
       throw;
    }
    app1->shutdown();
-}
+}
```
Comment (pmconrad) [inline tests/cli/main.cpp]: Please undo this change.

File: libraries/wallet/wallet.cpp [hunk 9]
```diff
@@ -770,6 +773,14 @@ class wallet_api_impl
 
       return true;
    }
+
+    void quit()
+    {
+        ilog( "Exiting Wallet ..." );
+
+        _Exit(0);
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: Like others have commented, calling exit() (or any of its variants) is bad style.

File: tests/cli/main.cpp (hunk 8 above)
Comment (cogutvalera) [reply tests/cli/main.cpp]: ok sure, thx !

File: libraries/wallet/wallet.cpp (hunk 9 above)
Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: yes we told about that a lot ! I will do soon ! Just need to find the best solution and approach.

**Comment by cogutvalera** [comment]
this PR depends also on this PR inside FC library https://github.com/bitshares/bitshares-fc/pull/63

**Comment by cogutvalera** [comment]
now without any exit method :-) just with cli::stop() command, is it OK now ? What are your thoughts ? Is this issue finished or not yet ? Should I do any changes or not ?

**Comment by pmconrad** [comment]
Note: need to bump fc after merging https://github.com/bitshares/bitshares-fc/pull/63

File: libraries/wallet/include/graphene/wallet/wallet.hpp [hunk 10]
```diff
@@ -632,6 +632,14 @@ class wallet_api
        */
       bool    load_wallet_file(string wallet_filename = "");
 
+      /** Exits from BitShares wallet.
+       * 
+       * The current wallet will be closed.
+       * 
+       * @returns SIGNAL
```
Comment (abitmore) [inline libraries/wallet/include/graphene/wallet/wallet.hpp]: Need to fix this comment.

File: libraries/wallet/wallet.cpp [hunk 11]
```diff
@@ -74,6 +74,9 @@
 #include <graphene/debug_witness/debug_api.hpp>
 #include <fc/smart_ref_impl.hpp>
 
+#include <cstdlib>
+using namespace std;
```
Comment (abitmore) [inline libraries/wallet/wallet.cpp]: Do we still need above lines?

File: tests/cli/main.cpp (hunk 8 above)
Comment (abitmore) [reply tests/cli/main.cpp]: Please undo this change (you removed the `newline` at end of file).

**Comment by cogutvalera** [comment]
ok sure !

File: libraries/wallet/include/graphene/wallet/wallet.hpp (hunk 10 above)
Comment (cogutvalera) [reply libraries/wallet/include/graphene/wallet/wallet.hpp]: ok sure ! sorry ! Thanks !

File: libraries/wallet/wallet.cpp (hunk 11 above)
Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: ok sure ! sorry ! Thanks ! forgot to remove those lines :-) I think we don't need these lines, but I will check to be 100% sure !

**Comment by cogutvalera** [comment]
rebased this PR already

**Comment by abitmore** [comment]
Check https://travis-ci.org/bitshares/bitshares-core/builds/406753086. Seems the `quit` command is triggering this issue: https://github.com/bitshares/bitshares-core/issues/177?

**Comment by abitmore** [comment]
Maybe not the case, see https://travis-ci.org/bitshares/bitshares-core/builds/406753086#L4778:
> 3422600ms th_a       wallet.cpp:780                quit                 ] Quiting Cli Wallet ...
Running 5 test cases...
/home/travis/build/bitshares/bitshares-core/tests/cli/main.cpp(244): error in "cli_quit": exception thrown by con.wallet_api_ptr->quit()

**Comment by cogutvalera** [comment]
I need to fix test code for quit command, sorry forget about test code after other changes on quit command.

File: libraries/wallet/wallet.cpp [hunk 12]
```diff
@@ -774,6 +774,14 @@ class wallet_api_impl
 
       return true;
    }
+
+    void quit()
+    {
+        ilog( "Quiting Cli Wallet ..." );
```
Comment (pmconrad) [inline libraries/wallet/wallet.cpp]: Typo: Quit*t*ing

**Comment by pmconrad** [comment]
I have merged https://github.com/bitshares/bitshares-fc/pull/63 - please bump fc to latest master.

File: libraries/wallet/wallet.cpp (hunk 12 above)
Comment (cogutvalera) [reply libraries/wallet/wallet.cpp]: sorry, my mistake :-) now will fix it !

**Comment by cogutvalera** [comment]
ok sure, thanks a lot ! now will do it ... sorry ...

**Comment by abitmore** [comment]
@cogutvalera the fc commit is not correct. Steps to bump FC:
```
cd bitshares-core
cd libraries
rm -rf fc
git submodule update --init --recursive
cd fc
git fetch origin
git checkout origin/master
git submodule update --init --recursive
cd ..
git add fc
git commit
```

**Comment by abitmore** [comment]
@cogutvalera: by the way, don't often rebase when in a PR, the reason is it will make review work harder (reviewers will have to read though all your code to find out what are the latest changes), similar to what I've mentioned here https://github.com/bitshares/bitshares-fc/pull/56#issuecomment-403509703
>@nanomobile just a note, in the future, please add new code with new commits but not always replace the old commit, to make it easier for others to know what you've newly changed, make the review work easier. Thanks.

**Comment by cogutvalera** [comment]
sorry ! understood now ! thought rebase is ok ! now I know what is the right way ! Thank you very much !

**Comment by abitmore** [comment]
@cogutvalera the FC version is still wrong. Please check again.

**Comment by cogutvalera** [comment]
sorry, my mistake with origin, because I've made all step by step as said in your "bump FC" instruction, but remote url name is core in my local workspace. I've replaced/fixed error, now hope all is correct ! Thanks a lot !

**Comment by abitmore** [comment]
This PR will close #1050. With the FC bump done in this PR, #929, #999 and #1127 can be closed as well.

**Comment by oxarbitrage** [comment]
removed my request for changes so you guys can merge.

**Comment by pmconrad** [comment]
Docker build failing due to problem with cli_wallet API documentation. Investigation. **Do not merge yet.**

**Comment by cogutvalera** [comment]
ok sure ! Thanks ! Will fix now !

**Comment by pmconrad** [comment]
Now it's wrong. void quit() doesn't return anything.
If you want to keep the comment you must fix the perl script. :-)

**Comment by cogutvalera** [comment]
ok understood :-) will remove it :-)

**Comment by abitmore** [comment]
@cogutvalera please bump FC to latest version to include the `time_point_sec` stringify test fix.

**Comment by cogutvalera** [comment]
Bump FC is done ! Thank you guys !

File: tests/cli/main.cpp [hunk 13]
```diff
@@ -221,6 +221,34 @@ BOOST_AUTO_TEST_CASE( cli_connect )
    app1->shutdown();
 }
 
+////////////////
+// Start a server and connect using the same calls as the CLI
+// Quit wallet and be sure that file was saved correctly
+////////////////
+BOOST_AUTO_TEST_CASE( cli_quit )
```
Comment (jmjatlanta) [inline tests/cli/main.cpp]: This test does not validate the wallet file was saved correctly.
1. Do we need to validate the wallet file? And if so...
2. What is the best way to validate it?

Comment (oxarbitrage) [reply tests/cli/main.cpp]: one idea i think i posted before for this will be something like:
- connect to wallet.
- create an account.
- quit.
- connect to wallet.
- check if created account is there. 

however it seems that the last version of the `quit` command is not saving the wallet.

Comment (cogutvalera) [reply tests/cli/main.cpp]: in cli_wallet::quit command we don't save file, should we save it ?

Comment (oxarbitrage) [reply tests/cli/main.cpp]: in my opinion yes but i am not following this issue very close in the last time, not sure if it was decided to just don't save.

Comment (cogutvalera) [reply tests/cli/main.cpp]: we need to decide what is the best way, save or not to save

Comment (jmjatlanta) [reply tests/cli/main.cpp]: Exiting cleanly should save the wallet file. Exiting with CTRL+D saves it. This should save it as well. I haven't looked, but I would imagine that the CTRL+D functionality would call a function, and the quit() functionality should call the same function.

Comment (abitmore) [reply tests/cli/main.cpp]: As discussed above, we need to save wallet file. The following code will save the wallet file: https://github.com/bitshares/bitshares-core/blob/4a2f333444b5d9a0cd13c78a16393594e9db1e6d/programs/cli_wallet/main.cpp#L296

Comment (abitmore) [reply tests/cli/main.cpp]: IMHO we can merge this PR and create a new issue about how to validate the wallet file.

Comment (cogutvalera) [reply tests/cli/main.cpp]: ok sure ! will do save wallet file and will improve test, soon will make commit !

Comment (cogutvalera) [reply tests/cli/main.cpp]: absolutely agree about new issue about validating how the wallet file was saved

**Comment by abitmore** [comment]
Manual test:
```
new >>> quit
quit
2551611ms th_a       wallet.cpp:780                quit                 ] Quitting Cli Wallet ...
9 canceled_exception: Canceled

2551611ms th_a       wallet.cpp:799                save_wallet_file     ] saving wallet to file wallet.json
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
```

* Best if `9 canceled_exception: Canceled` is not there, but this is a minor issue.
* wallet file does get saved
* the crash after saved wallet file is another common issue, if press Ctrl+D:

```
new >>>

2614126ms th_a       wallet.cpp:799                save_wallet_file     ] saving wallet to file wallet.json
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
```

File: tests/cli/main.cpp (hunk 13 above)
Comment (abitmore) [reply tests/cli/main.cpp]: Guys, please discuss the wallet file saving issue in https://github.com/bitshares/bitshares-core/issues/1109. Thanks.
